### PR TITLE
optimize, prioritize, and timebox processing of avatar updates

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* removed cruft
* prioritize avatar updates by proximity, position, and time since last update
* handle avatar updates by priority until out of time

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: c8741cb518d4db09614f882c16e187f5bc3262c4 available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-c8741cb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-c8741cb.dmg

**Comment by AndrewMeadows** [comment]
Below are some trace results for the main-thread.  The time spent in `OtherAvatars` is now capped at around 2msec which reduces Application::idle() times.  The total change for the 100 avatar test case in Application::idle() is from about 4.7 to 2.9 msec. Most of that is due to capping, but there are also a few optimization changes which help.

Master branch:
trace-main-thread-master

This PR:
trace-main-thread-sorted-avatar-updates-with-timebox

File: interface/src/avatar/Avatar.cpp [hunk 1]
```diff
@@ -127,6 +127,7 @@ Avatar::Avatar(RigPointer rig) :
     _nameRectGeometryID = geometryCache->allocateID();
     _leftPointerGeometryID = geometryCache->allocateID();
     _rightPointerGeometryID = geometryCache->allocateID();
+    _lastRenderUpdateTime = usecTimestampNow();
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: Avatar::_lastRenderUpdateTime is a new data member.  It is used in the prioritization calculation when sorting the avatar updates: avatars that have not been updated in a while will increase their priority until they eventually get updated, even in a starved situation where updates are capped after running over time budget.

File: interface/src/avatar/Avatar.cpp [hunk 2]
```diff
@@ -187,25 +188,35 @@ AABox Avatar::getBounds() const {
...
+        // TODO: rebuilding the shape constantly is somehwat expensive.
+        // We should only rebuild after significant change.
         rebuildCollisionShape();
     }
 }
 
+void Avatar::setTargetScale(float targetScale) {
+    AvatarData::setTargetScale(targetScale);
+    _isAnimatingScale = true;
+}
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: `setTargetScale()` moved out of base class `AvatarData` into derived class `Avatar`.

File: interface/src/avatar/Avatar.cpp [hunk 3]
```diff
@@ -302,80 +313,39 @@ void Avatar::updateAvatarEntities() {
...
-                qCDebug(interfaceapp) << "Optimizing" << (isMyAvatar() ? "myself" : getSessionUUID()) << "for visibility" << visibility;
-            }
-        }
-    }
-
-    uint64_t start = usecTimestampNow();
-    // CRUFT? _shouldSkipRender is never set 'true'
-    if (_shouldAnimate && avatarInView && !_shouldSkipRender) {
-        {
-            PerformanceTimer perfTimer("skeleton");
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: I removed this "culling" and "visibility" logic out of Avatar class.  The `AvatarManager` now computes a **priority** for each avatar and then processes them in order.  The "culling" behavior is now a fallout of how the prioritized updates are capped.  Visibility is also calculated elsewhere and is used to reduce the cost of `Avatar::simulate()`.  See `AvatarManager::updateOtherAvatars()` for details.

File: interface/src/avatar/Head.cpp [hunk 4]
```diff
@@ -117,7 +117,7 @@ void Head::simulate(float deltaTime, bool isMine, bool billboard) {
         }
     }
    
-    if (!(_isFaceTrackerConnected || billboard)) {
+    if (!_isFaceTrackerConnected) {
```
Comment (AndrewMeadows) [inline interface/src/avatar/Head.cpp]: The billboard logic for Head::simulate() was no longer used so I removed it.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3064dbbe6342a2a49215b86efd96d1998a1409c5 available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-3064dbb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-3064dbb.dmg

**Comment by AndrewMeadows** [comment]
What has changed:

We now time-box the avatar updates to 2msec.  After we've spent that budget we'll bail on Avatar updates until the next frame.  This may cause some avatars to momentarily freeze their animation, and maybe even freeze their position/orientation.  This will tend to be most visible for avatars that are not near the center of the view frustum (off to the far left or right).

To test:

(1) If you're using a computer with an i7 CPU it would be best to make sure this menu option is disabled:

Developer --> Render --> Enable Sparse Texture Management

This is the only way to get decent frame rates in the test region once everything has loaded however... it will cause the scene load to block and provide terrible frame rates for the first 15 or 20 seconds while textures load.

(2) Copy this simple one-line script to local file:

Window.location = "hifi://dev-chris.highfidelity.io/-11,3.0,42/0,-0.48175,0,0.87631";

(3) Find the interface.exe executable from this PR installer and run it from the command line like this:

./interface.exe --testScript C:/path/to/your/script/from/step/2.js

(4) Watch the scene load (there should be about 100 walking avatars in the destination region.  FPS will be terrible as they load but you should see them have a slight tendency to load from center of view toward the outer edges of the view.  Avatars to the side may freeze in place for a few seconds.

(5) Once the scene has loaded you should get better or equal frame rates as the current release in the same domain.

(6) Walk around with an eye on avatars that are near the sides of the view.  If any animation updates get dropped it will tend to be for distant avatars at the edges of the screen.  Verify that frame rate is not worse than for the official release in the same domain, and that most of the avatars continue to animate most of the time.

(7) Offer an opinion about how bad are the avatar animation artifacts, if any.

**Comment by AndrewMeadows** [comment]
test this please

File: interface/src/avatar/AvatarManager.cpp [hunk 5]
```diff
@@ -132,52 +132,131 @@ void AvatarManager::updateMyAvatar(float deltaTime) {
...
+            glm::vec3 avatarPosition = avatar->getPosition();
+            glm::vec3 offset = avatarPosition - frustumCenter;
+            float distance = glm::length(offset) + 0.001f; // add 1mm to avoid divide by zero
+            float radius = avatar->getBoundingRadius();
+            const glm::vec3& forward = cameraView.getDirection();
+            float apparentSize = radius / distance;
+            float cosineAngle = glm::length(offset - glm::dot(offset, forward) * forward) / distance;
+            const float TIME_PENALTY = 0.080f; // seconds
+            float age = (float)(startTime - avatar->getLastRenderUpdateTime()) / (float)(USECS_PER_SECOND) - TIME_PENALTY;
+            float priority = apparentSize + 0.25f * cosineAngle + age;
```
Comment (ZappoMan) [inline interface/src/avatar/AvatarManager.cpp]: coding standard: magic? what's the 0.25f here?

File: interface/src/avatar/AvatarManager.cpp [hunk 6]
```diff
@@ -132,52 +132,131 @@ void AvatarManager::updateMyAvatar(float deltaTime) {
...
+                _motionStatesToAddToPhysics.insert(motionState);
+                _motionStatesThatMightUpdate.insert(motionState);
+            }
+        }
+        avatar->animateScaleChanges(deltaTime);
+
+        uint64_t now = usecTimestampNow();
+        if (now < renderExpiry) {
+            // we're within budget
+            bool inView = sortData.priority > 0.5f * OUT_OF_VIEW_PENALTY;
```
Comment (ZappoMan) [inline interface/src/avatar/AvatarManager.cpp]: magic? why 1/2?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3064dbbe6342a2a49215b86efd96d1998a1409c5 available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-3064dbb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-3064dbb.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f024cef24490b601b314441c791da34635c57313 available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-f024cef.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-f024cef.dmg

**Comment by samcake** [comment]
I have noticed a bug when testing with hmd on my i5 in dev-chris2
if i get closeto an avatar, typically a meter away, it freezes and as i step back from it, a new instace of the same avatar appears and start animating.
my 2 cents: our heuristic rank that avatar last and doesn't update it for a while
and then something recreate it completely (meaning re add It to the Scene)
It seems that the former render items are staying dormant there (no animation)

**Comment by hifi-gustavo** [comment]
Build console output for commit: e6b315546786377a2767a55aaecd9a408f723f73 available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-e6b3155.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-e6b3155.dmg

**Comment by OmegaHeron** [comment]
Looks like test domain has moved on to a new proto version.  Was going to throw my 2 cents into this one too.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5783db910bc20098469d23fd59e856540fe8690e available at: https://gist.github.com/2ca461546f80ad8124bec181d956af44<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-5783db9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9493-5783db9.dmg

**Comment by AndrewMeadows** [comment]
@OmegaHeron I rebased this branch onto hifi/master so its version should now be compatible.

**Comment by OmegaHeron** [comment]
Thanks @AndrewMeadows, I'm enjoying the swarm with it now

**Comment by OmegaHeron** [comment]
Tested Windows 10, i5 6500 CPU, GTX970 GPU
QA Pass
Tested with desktop and DK2 HMD
General smoke test = pass

Specifically;
This PR offers a much smoother loading experience - less severe plummets in FPS and I didn't find any case while walking around in desktop or HMD mode where it seemed unnatural in how avatars were animating - be they edges of view or not.  Once loaded I found this PR to be smoother while walking around scene - no real FPS difference as I hold a steady 60 or 75 for desktop/DK2 in master and this PR, I suspect if I unlocked vsync I'd see a bit of FPS > in this PR vs master or, it's just gods of random.

No issue seen while walking around with my bubble on.

Seems good stuff to me.

**Pull request merged**